### ISSUE TITLE
device.enableSynchronization() doesn't work and messing with the other tests

### ISSUE THREAD

**Issue by hamidhadi** [wix/Detox#2630]
### Description
In some of my test cases, I have to disable auto synchronisation and wait for an element manually by `waitFor` API.

- [x] I have tested this issue on the latest Detox release and it still reproduces

#### Reproduction

```javascript
describe("my test suite", async () => {
     it("needs to handle synchronisation manually", async () => {
        await device.disableSynchronization();
        // Doing some stuff
        await waitFor(element(by.id("myElement"))).toBeVisible()
         
        await device.enableSynchronization();
     })

     it("needs to handle synchronisation automatically", async () => {
        // All my tests here is failing because auto synchronisation is disabled ⛔ 
     })
})
```

But if I relaunch the app, it works.

```javascript
describe("my test suite", async () => {
     it("needs to handle synchronisation manually", async () => {
        await device.disableSynchronization();
        // Doing some stuff
        await waitFor(element(by.id("myElement"))).toBeVisible()
         
        await device.launch({ newInstance: true });
     })

     it("needs to handle synchronisation automatically", async () => {
        // Everything is fine here ✅ 
     })
})
```

#### Expected behavior
The `await device.enableSynchronization();` should work as expected and prevent messing with the other test cases.

#### Environment (please complete the following information):
 - Detox:  `18.2.2`
 - React Native:  `0.63.2`
 - Node:  `15.5.1`
 - Device:  `Simulator iPhone 11Pro`
 - Xcode:  `12.4`
 - iOS:  `14.4`
 - macOS:   `11.1`

#### If you are experiencing a timeout in your test

- [x] I have followed the instructions under [Identifying which synchronization mechanism causes us to wait too much](https://github.com/wix/Detox/blob/master/docs/Troubleshooting.Synchronization.md#identifying-which-synchronization-mechanism-causes-us-to-wait-too-much), I have read [synchronization debug documentation](https://github.com/wix/DetoxSync/blob/master/IdleStatusDocumentation.md) and am providing the relevant synchronization debug output below:

**Comment by LeoNatan** [comment]
@alon-ha Please take a look.

**Comment by alon-ha** [comment]
@hamidhadi I tried to reproduce but the issue does not reproduce for me.  
`await device.enableSynchronization()` works as expected in the example I tried.
Provide a demo project were a test (another `it`) is failing after the sequence of disable/enable synchronization

**Comment by denissb** [comment]
@hamidhadi  Are you running the tests in parallel with `--workers 2`?

For me when I run the tests in parallel it fails similarly if I disable synchronisation. Apparently the other test running in parallel fails because it needs synchronisation but it gets turned off for all instances. I came to the conclusion that one should either find a way to fix the test to run with synchronisation enabled or you just run one test at a time.

**Comment by dani-mp** [comment]
@denissb Hey! I work w/ @hamidhadi. Yes, we're running tests in parallel. What you say makes sense, but I guess the ideal would be that disabling sync only does it for a concrete instance. We make sure that we disable sync at the last moment possible and enable it again afterward, but we can't run that test with sync enabled right now, and losing the power of `--workers 2` wouldn't be great either.

Let's see if @alon-ha has any ideas!

**Comment by d4vidi** [comment]
Disabling sync does *not* does that in all running instances, period. This is not the direction towards solving this issue...

**Comment by hamidhadi** [comment]
@alon-ha Sorry for being late. 
Here is an example of the scenario that we're facing. Please let me know if there is something unclear.
https://github.com/hamidhadi/DetoxSynchronisation.

**Comment by dani-mp** [comment]
Hey, @alon-ha & @d4vidi. We put together the project showcasing the issue, do you think you have time to take a look?

Otherwise, do you prefer we create a separate issue for this?

Thanks a lot for your help.

**Comment by d4vidi** [comment]
@dani-mp Thanks a million - we will take look, no need for a separate issue 🙏🏻
We are currently on a holiday break, I hope to get to this soon enough.

**Comment by dani-mp** [comment]
@d4vidi enjoy your holiday!

We did use Detox 17 and I believe we had the same issue, but I'm not 100 % sure. Then we transitioned to 18 to see if it was fixing it. We went back to 17 for some time because of some problems we had in CI (some extra flakiness), but we understood that it wasn't caused by the version and moved back to 18.

**Comment by d4vidi** [comment]
@dani-mp in the meantime, do note that the construct:
```
await waitFor(element(by.id("myElement"))).toBeVisible()
```
is scarce.

You need to apply a termination command, such as `withTimeout()` in order for it to have an effect:

```
await waitFor(element(by.id("myElement"))).toBeVisible().withTimeout(1000)
```

**Comment by dani-mp** [comment]
@d4vidi yup, we're now aware of this. I think the sample provided respects this, it's just the original description of the issue what is wrong.

**Comment by dani-mp** [comment]
Hey, @d4vidi! Could you confirm the issue or was it us doing something wrong?

**Comment by d4vidi** [comment]
@alon-ha and I will take a look today

**Comment by d4vidi** [comment]
@dani-mp Looking at the example project, seems that the very call to `device.enableSynchronization()` triggers a wait until the app goes idle. *This is what's blocking your tests* -- not the various assertions. 
Whether that's good practice or not, is definitely something to discuss internally. We will "take it offline" and think that through (namely, whether `enableSynchronization()` should just enable and return). You are more than welcome to solicit your opinion! 🙏🏻 

In any case, we value your patience and work a great deal!

**Comment by d4vidi** [comment]
We've decided to keep as-is (actually, it is not really possible to do otherwise). [This](https://github.com/wix/Detox/commit/dd1d5c7deedf379aecf9841e00db80157e83ba99) closes this issue.

**Comment by dani-mp** [comment]
@d4vidi thanks, folks, for the update and clarifying it on the docs!
